### PR TITLE
create binary, Makefile and basic build/test workflow

### PR DESCRIPTION
**Pull request by skriss**
Adds the envoy-gateway binary, a Makefile
and a basic GitHub Actions workflow for
running build and test jobs on each PR & push.

Signed-off-by: Steve Kriss <krisss@vmware.com>

### CODE REVIEW

**Comment by skriss** [comment]
There may be a repo setting related to GH Actions that needs to change to allow actions to run on PRs from forks? I don't have access to change.

**Comment by danehans** [comment]
@mattklein123 PTAL at https://github.com/envoyproxy/gateway/pull/72#issuecomment-1137454764. Can you confirm the GH actions repo [settings](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository)?

File: .github/workflows/build_and_test.yaml
Comment (danehans) [inline .github/workflows/build_and_test.yaml:20]: Can an [env var](https://docs.github.com/en/actions/learn-github-actions/environment-variables) be used so we only have to update the go version in 1 place?

**Review by danehans — CHANGES_REQUESTED** [review]
@skriss thanks for taking the lead on this. I have 1 small nit.

File: .github/workflows/build_and_test.yaml
Comment (skriss) [reply .github/workflows/build_and_test.yaml]: :+1: done

**Review by danehans — DISMISSED** [review]
LGTM

**Comment by mattklein123** [comment]
It's set to require approval for first time contributors, but otherwise should run from forks.

File: Makefile
Comment (arkodg) [inline Makefile:3]: we can deal with cross compilation and LD flags later but we should at least save the binary in a separate dir such as `bin` or `build`

Comment (arkodg) [inline Makefile:3]: we can deal with cross compilation and LD flags later but we should at least save the binary in a separate dir such as `bin` or `build`

File: internal/cmd/root.go
Comment (arkodg) [inline internal/cmd/root.go:25]: `envoy-gateway` is a lot to type, but I can't think of anything better rn,
will the server/daemon CLI (`envoy-gateway`) be different that the client CLI ( which I envision will be required for debuggability) or will they be subcommands within `envoy-gateway`  (I'd vote for the latter) ?

Comment (arkodg) [inline internal/cmd/root.go:25]: `envoy-gateway` is a lot to type, but I can't think of anything better rn,
will the server/daemon CLI (`envoy-gateway`) be different that the client CLI ( which I envision will be required for debuggability) or will they be subcommands within `envoy-gateway`  (I'd vote for the latter) ?

File: Makefile
Comment (skriss) [reply Makefile]: Works for me, moved to under `bin/`. Agreed on adding additional additional functionality here later, wanted to get started with the basics.

File: internal/cmd/root.go
Comment (skriss) [reply internal/cmd/root.go:25]: Yeah, `envoy-gateway` seemed better than `gateway`, but it can certainly change to something else if we're not happy with it.

Re: CLI, I'm not sure, I don't think we've discussed that much, but if it's mainly for debugging I think it's fine to use subcommands on the same binary (that's what we do in Contour, FWIW).

File: cmd/envoy-gateway/main.go
Comment (arkodg) [inline cmd/envoy-gateway/main.go:26]: lets decide on a logger and/or logging framework either here or via a GH issue
I'd recommend using a generic logging framework thats the project is based on such as https://github.com/tetratelabs/log#common-logging-infrastructure which allows us to pick any logging implementation and swap it out in the future
cc @basvanbeek

Comment (skriss) [reply cmd/envoy-gateway/main.go:26]: Yep, the choice of logging framework hadn't been resolved in the open doc so I didn't want to touch it yet, I'd suggest deferring that to a future PR.

File: .github/workflows/build_and_test.yaml
Comment (lizan) [inline .github/workflows/build_and_test.yaml:12]: Use `cache` from `setup-go`?
https://github.com/actions/setup-go#caching-dependency-files-and-build-outputs

Comment (skriss) [reply .github/workflows/build_and_test.yaml:12]: Well, it hadn't been released as of yesterday when I did this, but now it has: https://github.com/actions/setup-go/releases/tag/v3.2.0

I'll try it out.

Comment (skriss) [reply .github/workflows/build_and_test.yaml:12]: Looked good in testing but I'm going to hold off on updating until the `v3` tag is updated to point to this latest setup-go release, so we're not pinned to a specific version of the action and automatically get any patch releases.

File: .github/workflows/build_and_test.yaml
Comment (arkodg) [inline .github/workflows/build_and_test.yaml:3]: lets limit CI to only run on `main` and not run when `.md` files or images are updated
```
  push:
    branches:
      - main
    paths-ignore:
      - "**/*.md"
      - "**/*.png"
  pull_request:
    branches:
      - main
    paths-ignore:
      - "**/*.md"
      - "**/*.png"
```

File: cmd/envoy-gateway/main.go
Comment (danehans) [reply cmd/envoy-gateway/main.go:26]: I created https://github.com/envoyproxy/gateway/issues/76 to track the logging framework requirement. I suggest that #76 does not block this PR.

File: .github/workflows/build_and_test.yaml
Comment (skriss) [reply .github/workflows/build_and_test.yaml:3]: I'll make the path change; re: branches, we'll have release branches that we want this to run on (i.e. we already have https://github.com/envoyproxy/gateway/tree/release-v0.1.0). Let's leave it running on all branches for now, if we find we have branches that we *need* it not to run on, we can always add later. CI minutes are free anyway :)

File: .github/workflows/build_and_test.yaml
Comment (skriss) [reply .github/workflows/build_and_test.yaml]: Alright, they've updated the v3 tag so made this change.

File: .github/workflows/build_and_test.yaml
Comment (danehans) [reply .github/workflows/build_and_test.yaml:3]: +1

**Review by danehans — APPROVED** [review]
LGTM

**Review by lizan — APPROVED** [review]

File: .github/workflows/build_and_test.yaml
Comment (lizan) [inline .github/workflows/build_and_test.yaml:1]: emm, actually this file need to be named `build_and_test.yml` for GitHub to pick it up.

**Review by lizan — CHANGES_REQUESTED** [review]

File: .github/workflows/build_and_test.yaml
Comment (skriss) [reply .github/workflows/build_and_test.yaml:1]: It should be fine as-is, it's running correctly on my branch (https://github.com/skriss/gateway/actions/runs/2392566235) as well as on a test PR I created in my fork (https://github.com/skriss/gateway/pull/1/checks). Most of our workflow files in Contour also use the `.yaml` extension.  I'm still not sure why it's not running on this PR, but maybe we can merge & see if it runs on `main` and on the next PR.

Comment (lizan) [reply .github/workflows/build_and_test.yaml:1]: OK let's give it a try as is.

**Review by lizan — APPROVED** [review]

**Comment by alexgervais** [comment]
🎉

**Comment by skriss** [comment]
All looks good:
- push to main (https://github.com/envoyproxy/gateway/actions/runs/2394494530)
- new PR with changes to .go file (https://github.com/envoyproxy/gateway/actions/runs/2397267584)

**Pull request merged**
